packages/core/contracts/UnirepSocial.sol

```diff
@@ -265,4 +269,39 @@ contract UnirepSocial {
     ) external {
         unirep.updateUserStateRoot(publicSignals, proof);
     }
+
+     /*
+     * Set new user name for an epochKey
+     * @param epochKey epoch key that attempts to set a new uername
+     * @param oldUsername oldusername that the eppch key previously claimed
+     * @param newUsername requested new user name
+     */
+     function setUsername(
+        uint256 epochKey,
+        uint256 oldUsername,
+        uint256 newUsername
+     ) external payable {
+        uint attestingFee = unirep.attestingFee();
+
+        // check if the new username is not taken
+        require(usernames[newUsername] == false, "This username is already taken");
+
+        // only admin can call this function
+        require(msg.sender == admin, "Only admin can send transactions to this contract");
+
+        usernames[oldUsername] = false;
+        usernames[newUsername] = true;
+
+        // attest to the epoch key to give the key the username
+        Unirep.Attestation memory attestation;
+        attestation.attesterId = attesterId;
+        attestation.posRep = 0;
+        attestation.negRep = 0;
+        attestation.graffiti = newUsername;
```

Comment:
The `username` should be hashed by a `Poseidon` hash function
so in the reputation proof the user can prove the username is the pre-image of the graffiti.

Reply:
I am not sure whether it will be included in the PR because the `Poseidon` library should be linked as well.

Reply:
I was planning to hash it off chain so we don’t have to worry about that.